### PR TITLE
Add Swift modulemap for WebKit Private APIs

### PR DESCRIPTION
**Pull request by ian-twilightcoder**
#### db17513feb3f72a858f82a6bd062ba5127071795
<pre>
Add Swift modulemap for WebKit Private APIs

<a href="https://bugs.webkit.org/show_bug.cgi?id=203059">https://bugs.webkit.org/show_bug.cgi?id=203059</a>
rdar://37471861

Reviewed by Elliott Williams and Dean Jackson.

Add private module maps for WebKit. WebEventRegion.h is conditionally installed,
so the private module maps need to be unifdef&apos;ed and manually copied rather than
using the normal Xcode mechanism.
DerivedSources.make isn&apos;t passing an SDK or -target to clang to figure out
FEATURE_AND_PLATFORM_DEFINES because they aren&apos;t defined yet, fix that by
defining those earlier.
Fix all of the private headers to compile in Objective-C, mostly by adding C++
guards.

* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Modules/OSX_Private.modulemap: Added.
* Source/WebKit/Modules/iOS_Private.modulemap: Added.
* Source/WebKit/Platform/cocoa/WKCrashReporter.h:
* Source/WebKit/Platform/unix/EnvironmentUtilities.h:
* Source/WebKit/Shared/Cocoa/XPCEndpoint.h:
* Source/WebKit/Shared/Cocoa/XPCEndpointClient.h:
* Source/WebKit/Shared/WebPushDaemonConstants.h:
* Source/WebKit/UIProcess/API/C/WKNotificationManager.h:
* Source/WebKit/UIProcess/API/C/WKWebsitePolicies.h:
* Source/WebKit/UIProcess/API/Cocoa/WKContextMenuElementInfoPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKRemoteWebInspectorViewControllerPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanelForTesting.h:
* Source/WebKit/UIProcess/API/cpp/WKRetainPtr.h:
* Source/WebKit/UIProcess/ios/fullscreen/FullscreenTouchSecheuristic.h:
* Source/WebKit/UIProcess/ios/fullscreen/FullscreenTouchSecheuristicParameters.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/262032@main">https://commits.webkit.org/262032@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/8052d5232e7a919ea7403157231cfc30e50b13ed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/308 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/324 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/337 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/314 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/292 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/347 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/544 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/314 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/264 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/305 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/339 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/287 "Build was cancelled. Recent messages:OS: Monterey (12.6), Xcode: 14.0; Skipping applying patch since patch_id isn't provided; Checked out pull request") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/298 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/369 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/274 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/315 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/297 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/267 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/307 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/78 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/300 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by aproskuryakov** [comment]
What does this mean for Objective-C++ clients of WebKit, will they still be able to use C++ from our private headers?

**Comment by ian-twilightcoder** [comment]
> What does this mean for Objective-C++ clients of WebKit, will they still be able to use C++ from our private headers?

Objective-C++ clients should be unaffected, `__cplusplus` is defined so they'll go ahead like normal.

**Comment by emw-apple** [comment]
Is there anything we can do to avoid committing a bespoke list of private headers in every modulemap? Seems like these will be a pain to maintain.

**Comment by ian-twilightcoder** [comment]
> Is there anything we can do to avoid committing a bespoke list of private headers in every modulemap? Seems like these will be a pain to maintain.

It actually should be pretty easy to maintain, if you delete a header the module verifier will complain that the module map has a missing header, if you add one it'll complain that the module map didn't get it added too.

However, the other thing we can do is exclude WebKitPrivate.h from the private module, and then we can do PrivateHeaders as the umbrella directory and `explicit module * { export * }` instead. Are clients generally intended to use WebKitPrivate.h, or is that old cruft that we might even want to delete?

**Comment by ian-twilightcoder** [comment]
I can't seem to figure this one out in the iOS failure
```
/Volumes/Data/worker/iOS-16-Build-EWS/build/WebKitBuild/Release-iphoneos/WebKit.framework/Modules/module.private.modulemap:2183:12: error: header 'WebEventRegion.h' not found
    header "WebEventRegion.h"
           ^
```
As far as I can see, that header is installed in PrivateHeaders. MobileMiniBrowser isn't somehow building before WebKit finishes is it?

**Comment by emw-apple** [comment]
WebEventRegion.h is only copied when building with an internal SDK (corresponding to the `ENABLE(IOS_TOUCH_EVENTS)` platform macro). Here's the inclusion logic: https://github.com/WebKit/WebKit/blob/cc4248a1c61e48abca0a20cce40db5f416b075c5/Source/WebKit/Configurations/WebKit.xcconfig#L222-L227

**Comment by ian-twilightcoder** [comment]
> WebEventRegion.h is only copied when building with an internal SDK (corresponding to the `ENABLE(IOS_TOUCH_EVENTS)` platform macro). Here's the inclusion logic:
> 
> https://github.com/WebKit/WebKit/blob/cc4248a1c61e48abca0a20cce40db5f416b075c5/Source/WebKit/Configurations/WebKit.xcconfig#L222-L227

Ahhh, thanks. That's going to make the module map (more) complicated... I think we're going to have to `unifdef` it

**Comment by ian-twilightcoder** [comment]
> > Is there anything we can do to avoid committing a bespoke list of private headers in every modulemap? Seems like these will be a pain to maintain.
> 
> It actually should be pretty easy to maintain, if you delete a header the module verifier will complain that the module map has a missing header, if you add one it'll complain that the module map didn't get it added too.
> 
> However, the other thing we can do is exclude WebKitPrivate.h from the private module, and then we can do PrivateHeaders as the umbrella directory and `explicit module * { export * }` instead. Are clients generally intended to use WebKitPrivate.h, or is that old cruft that we might even want to delete?

From the look of it though, WebKitPrivate.h looks like it has just new(er) "WebKit 2" headers, so that would suggest that it is meant to be used as a general private umbrella.

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj
Comment (ian-twilightcoder) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj]: I'm not sure if you want this pulled into Scripts instead. It probably has some style issues, though I tried to mostly match how some of the other scripts look.
```
HAS_WEBEVENTREGION=0
for includedSourceFileName in $INCLUDED_SOURCE_FILE_NAMES
do
    if [ "$includedSourceFileName" = "WebEventRegion.h" ]
    then
        HAS_WEBEVENTREGION=1
    fi
done

if [ $HAS_WEBEVENTREGION -eq 0 ]
then
    moduleMapUnifdefFlags="-UHAS_WEBEVENTREGION"
else
    moduleMapUnifdefFlags="-DHAS_WEBEVENTREGION"
fi

unifdef $moduleMapUnifdefFlags -o "$SCRIPT_OUTPUT_FILE_0" "$SCRIPT_INPUT_FILE_0"

returnValue=$?
if [ \( $returnValue -ne 0 \) -a \( $returnValue -ne 1 \) ]
then
    exit $returnValue
fi
```

Comment (grorg) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: It probably is better as a separate script, but I'm not sure how important this is.

**Comment by grorg** [comment]
This looks ok to me. I'll mark it as reviewed, but I think @emw-apple and @aproskuryakov should probably give their opinion.

**Review by grorg — APPROVED** [review]

**Comment by aproskuryakov** [comment]
> This looks ok to me. I'll mark it as reviewed, but I think @emw-apple and @aproskuryakov should probably give their opinion.

Mostly @emw-apple. As far as I'm concerned, I just want to confirm that you've successfully built Safari for macOS, iOS and Safari Technology Preview with this change applied. Please reach out internally if there are any challenges with trying that out.

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj
Comment (emw-apple) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: I think it would be easier / more in line with project conventions to do this in WebKit's `DerivedSources.make`. From inside that Makefile, we have access to platform macros like `ENABLE(IOS_TOUCH_EVENTS)`, so we don't have to piggy-back off the INCLUDED_SOURCE_FILE_NAMES variable, and WebKit developers will better recognize the platform flag we're basing this off of.

Plus, it would mean we can use the normal MODULEMAP_PRIVATE_FILE build setting, instead of having to install the modulemap into the right place ourselves.

I played around a bit and came up with:
```makefile
PRIVATE_MODULEMAPS = \
	Modules/iOS_Private.modulemap \
	Modules/MacCatalyst_Private.modulemap \
	Modules/OSX_Private.modulemap \
#

GENERATED_MODULEMAPS = $(notdir $(PRIVATE_MODULEMAPS))

%.modulemap : Modules/%.modulemap
	unifdef -B $(addprefix -D, $(FEATURE_AND_PLATFORM_DEFINES)) -o $@ $<

all : $(GENERATED_MODULEMAPS)
```

This generates files like:
```
WebKitBuild/Debug/DerivedSources/WebKit/MacCatalyst_Private.modulemap
WebKitBuild/Debug/DerivedSources/WebKit/OSX_Private.modulemap
WebKitBuild/Debug/DerivedSources/WebKit/iOS_Private.modulemap
```

which can be accessed using `$(BUILT_PRODUCTS_DIR)/DerivedSources` in xcconfig.

If you do this, the conditional block in the modulemap should be:
```
#ifdef ENABLE_IOS_TOUCH_EVENTS
  explicit module WebEventRegion {
    header "WebEventRegion.h"
    export *
  }
#endif
```

Comment (ian-twilightcoder) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: Spiffy, let me give that a try

**Comment by emw-apple** [comment]
> > > Is there anything we can do to avoid committing a bespoke list of private headers in every modulemap? Seems like these will be a pain to maintain.
> > 
> > 
> > It actually should be pretty easy to maintain, if you delete a header the module verifier will complain that the module map has a missing header, if you add one it'll complain that the module map didn't get it added too.

OK, sweet! That's fine with me, thanks for clarifying.

> > However, the other thing we can do is exclude WebKitPrivate.h from the private module, and then we can do PrivateHeaders as the umbrella directory and `explicit module * { export * }` instead. Are clients generally intended to use WebKitPrivate.h, or is that old cruft that we might even want to delete?
> 
> From the look of it though, WebKitPrivate.h looks like it has just new(er) "WebKit 2" headers, so that would suggest that it is meant to be used as a general private umbrella.

Yeah, I think it is intended to be an umbrella header for WebKit SPI. Maybe we don't need it and could just export everything from PrivateHeaders, but I think it would take additional work, and it might break existing clients.

I wouldn't be surprised if we have some things in private headers that really shouldn't be available to clients — for example, we sometimes stuff testing-only headers and scripts into private headers so that other parts of the WebKit project can consume them.

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj
Comment (ian-twilightcoder) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: Xcode requires that the module map files already be present at the start of a build, so we still can't use `MODULEMAP_PRIVATE_FILE`. Is there a reasonable way for DerivedSources to install the module map files in the framework?

Comment (emw-apple) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: Aw, too bad. Best I can think of is:

- Add the modulemap to WebKit's "Derived Sources" group. Easiest way to do this is to generate it once, and then use `Add Files to "WebKit"…` on the group.
- Create a Copy Files phase that copies it to "Wrapper", subpath `$(WK_FRAMEWORK_VERSION_PREFIX)Modules` (or MODULES_FOLDER_PATH, but not sure if that works in this context)

Comment (emw-apple) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: We'd probably need to have Make name the output file `module.private.modulemap` in all cases, so like:
```make
ifneq ($(findstring iphone,$(WK_PLATFORM_NAME)),)
module.private.modulemap : Modules/iOS_Private.modulemap
else ifeq (maccatalyst,$(WK_PLATFORM_NAME))
module.private.modulemap : Modules/MacCatalyst_Private.modulemap
else ifeq (macosx,$(WK_PLATFORM_NAME))
module.private.modulemap : Modules/OSX_Private.modulemap
endif

%.modulemap : 
	unifdef -B $(addprefix -D, $(FEATURE_AND_PLATFORM_DEFINES)) -o $@ $<

all : module.private.modulemap
```

Comment (ian-twilightcoder) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: 👍 let me give that a spin

Comment (ian-twilightcoder) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: You can kind of do "Products Directory" + `MODULES_FOLDER_PATH`, but it's technically wrong since you really want `$(TARGET_BUILD_DIR)/$(MODULES_FOLDER_PATH)` and "Products Directory" + `MODULES_FOLDER_PATH` is actually `$(BUILT_PRODUCTS_DIR)/$(MODULES_FOLDER_PATH)` So "Wrapper" + `$(WK_FRAMEWORK_VERSION_PREFIX)Modules` is regrettably what we need to do ("Wrapper" is `$(TARGET_BUILD_DIR)/$(WRAPPER_NAME)`)

File: Source/WebKit/Configurations/WebKit.xcconfig
Comment (emw-apple) [inline Source/WebKit/Configurations/WebKit.xcconfig:242]: There's a non-zero chance that this will impact production builds, but if testing shows it working fine, then LGTM. (cc @aproskuryakov)

**Review by emw-apple — APPROVED** [review]
DerivedSource + Copy Files phase looks great. Thank you!

File: Source/WebKit/Configurations/WebKit.xcconfig
Comment (aproskuryakov) [reply Source/WebKit/Configurations/WebKit.xcconfig:242]: Should probably diff an installhdrs root to verify what changes.

Comment (ian-twilightcoder) [reply Source/WebKit/Configurations/WebKit.xcconfig:242]: Yeah, I'll do that.

**Comment by ian-twilightcoder** [comment]
> DerivedSource + Copy Files phase looks great. Thank you!

It doesn't work though, `FEATURE_AND_PLATFORM_DEFINES` only collects the true defines, it doesn't get the false ones. Trying to fix that.

File: Source/WebKit/DerivedSources.make
Comment (emw-apple) [inline Source/WebKit/DerivedSources.make]: This should do it in one compiler invocation:

```make
FEATURE_AND_PLATFORM_FLAGS := $(shell $(call platform_h_compiler_command,-E -P -dM) | $(PERL) -ne "print if s/\#define ((HAVE_|USE_|ENABLE_|WTF_PLATFORM_)\w+) (\d)/\1=\2/")
FEATURE_AND_PLATFORM_DEFINES := $(patsubst %=1, %, $(FEATURE_AND_PLATFORM_FLAGS))
FEATURE_AND_PLATFORM_UNDEFINES := $(patsubst %=0, %, $(FEATURE_AND_PLATFORM_FLAGS))
```

**Review by emw-apple — COMMENTED** [review]
> It doesn't work though, `FEATURE_AND_PLATFORM_DEFINES` only collects the true defines, it doesn't get the false ones. Trying to fix that.

Ah, of course, unifdef isn't a real preprocessor so we can't tell it to simply "remove any undefined macro".

File: Source/WebKit/DerivedSources.make
Comment (ian-twilightcoder) [reply Source/WebKit/DerivedSources.make]: Aha, thanks. My make skills are... rudimentary 🙄

Comment (emw-apple) [reply Source/WebKit/DerivedSources.make]: Consider yourself lucky!

**Comment by ian-twilightcoder** [comment]
@emw-apple would you mind taking one more look at this one? Just for the new DerivedSources.xcconfig, and the few tweaks in the DerivedSources makefile and input/output lists. Thanks!

File: Source/WebKit/Configurations/DerivedSources.xcconfig
Comment (emw-apple) [inline Source/WebKit/Configurations/DerivedSources.xcconfig]: So in other projects (like WebCore), instead of creating a separate set of configs for the Derived Sources target, we use the main framework's xcconfig to get settings like INSTALLHDRS_SCRIPT_PHASE. Would you mind doing that here as well? It would also mean going back to having the modulemap input path logic computed in the Makefile.

Here's the diff to change WebKit's Derived Sources to be based off of WebKit.xcconfig:
```diff
diff --git a/Source/WebKit/WebKit.xcodeproj/project.pbxproj b/Source/WebKit/WebKit.xcodeproj/project.pbxproj
index 122fc59c6720..748aa6389e08 100644
--- a/Source/WebKit/WebKit.xcodeproj/project.pbxproj
+++ b/Source/WebKit/WebKit.xcodeproj/project.pbxproj
@@ -18987,7 +18987,7 @@
                };
                C0CE72861247E66800BC0EC4 /* Debug */ = {
                        isa = XCBuildConfiguration;
-                       baseConfigurationReference = 5DAD73F1116FF90C00EE5396 /* BaseTarget.xcconfig */;
+                       baseConfigurationReference = BCB86F4B116AAACD00CE20B7 /* WebKit.xcconfig */;
                        buildSettings = {
                                COPY_PHASE_STRIP = NO;
                                GCC_DYNAMIC_NO_PIC = NO;
@@ -18998,7 +18998,7 @@
                };
                C0CE72871247E66800BC0EC4 /* Release */ = {
                        isa = XCBuildConfiguration;
-                       baseConfigurationReference = 5DAD73F1116FF90C00EE5396 /* BaseTarget.xcconfig */;
+                       baseConfigurationReference = BCB86F4B116AAACD00CE20B7 /* WebKit.xcconfig */;
                        buildSettings = {
                                COPY_PHASE_STRIP = YES;
                                DEBUG_INFORMATION_FORMAT = "dwarf-with-dsym";
@@ -19008,7 +19008,7 @@
                };
                C0CE72881247E66800BC0EC4 /* Production */ = {
                        isa = XCBuildConfiguration;
-                       baseConfigurationReference = 5DAD73F1116FF90C00EE5396 /* BaseTarget.xcconfig */;
+                       baseConfigurationReference = BCB86F4B116AAACD00CE20B7 /* WebKit.xcconfig */;
                        buildSettings = {
                                PRODUCT_NAME = "Derived Sources";
                        };
```

Comment (emw-apple) [reply Source/WebKit/Configurations/DerivedSources.xcconfig]: If there's a good reason to have a separate DerivedSources.xcconfig that's escaping me, feel free to go ahead with the merge :)

Comment (ian-twilightcoder) [reply Source/WebKit/Configurations/DerivedSources.xcconfig]: Sure we can use the same xcconfig. We could still make a build setting for the private module map so that the xcfilelist can just have the platform appropriate module map. That way it won’t rebuild e.g. iOS if the macOS module map changes

Comment (emw-apple) [reply Source/WebKit/Configurations/DerivedSources.xcconfig]: > That way it won’t rebuild e.g. iOS if the macOS module map changes

Well, iirc it will rerun Make regardless becuase DerivedSources-input.xcconfig is platform-agnostic. But Make will avoid regenerating the private module map based on its own dependency checking.

Comment (ian-twilightcoder) [reply Source/WebKit/Configurations/DerivedSources.xcconfig]: That’s true. Mostly I just like the module map files being declared next to each other in the xcconfig. It’ll also avoid the Run Script that runs make just for a no-op, but that’s really not a big deal like you say. However you prefer 😊

Comment (emw-apple) [reply Source/WebKit/Configurations/DerivedSources.xcconfig]: > Mostly I just like the module map files being declared next to each other in the xcconfig.

That makes sense, putting the declarations side by side works for me!

Comment (ian-twilightcoder) [reply Source/WebKit/Configurations/DerivedSources.xcconfig]: Ok, I think I got it this time. Also I noticed that the Mac Catalyst module map is the same as the iOS one so I added one more conditional and combined them. Can I get one more review please?

- deleted MacCatalyst_Private.modulemap
- deleted DerivedSources.xcconfig and switched that target to using the WebKit config
- moved the private module map settings to WebKit.xcconfig
- added WK_IS_CATALYST_FLAG to DerivedSources.make

File: Source/WebKit/DerivedSources-input.xcfilelist
Comment (emw-apple) [inline Source/WebKit/DerivedSources-input.xcfilelist]: Does this work? I would have thought that the generator would insist on having separate `$(PROJECT_DIR)/Modules/…` paths, because it only does a few hard-coded variable substitutions.

File: Source/WebKit/Modules/iOS_Private.modulemap
Comment (emw-apple) [inline Source/WebKit/Modules/iOS_Private.modulemap]: What does WK_IS_CATALYST accomplish? In wtf/PlatformEnableCocoa.h, ENABLE_IOS_TOUCH_EVENTS is defined as `PLATFORM(IOS_FAMILY) && USE(APPLE_INTERNAL_SDK)`, and IOS_FAMILY includes Catalyst.

File: Source/WebKit/DerivedSources-input.xcfilelist
Comment (emw-apple) [reply Source/WebKit/DerivedSources-input.xcfilelist]: Yeah, I just ran a build locally and got:
```diff
diff --git a/Source/WebKit/DerivedSources-input.xcfilelist b/Source/WebKit/DerivedSources-input.xcfilelist
index 6e57f3b3df81..c9a80ed71b64 100644
--- a/Source/WebKit/DerivedSources-input.xcfilelist
+++ b/Source/WebKit/DerivedSources-input.xcfilelist
@@ -98,6 +98,7 @@ $(PROJECT_DIR)/GPUProcess/webrtc/RemoteMediaRecorder.messages.in
 $(PROJECT_DIR)/GPUProcess/webrtc/RemoteMediaRecorderManager.messages.in
 $(PROJECT_DIR)/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.messages.in
 $(PROJECT_DIR)/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.messages.in
+$(PROJECT_DIR)/Modules/OSX_Private.modulemap
 $(PROJECT_DIR)/NetworkProcess/Cookies/WebCookieManager.messages.in
 $(PROJECT_DIR)/NetworkProcess/CustomProtocols/LegacyCustomProtocolManager.messages.in
 $(PROJECT_DIR)/NetworkProcess/NetworkBroadcastChannelRegistry.messages.in
```

We'll need the real paths :\

File: Source/WebKit/Modules/iOS_Private.modulemap
Comment (ian-twilightcoder) [reply Source/WebKit/Modules/iOS_Private.modulemap]: Hmm, I'm not sure, but an Apple internal build somehow had `ENABLE_IOS_TOUCH_EVENTS =0` for Mac Catalyst. 🤔

File: Source/WebKit/DerivedSources-input.xcfilelist
Comment (ian-twilightcoder) [reply Source/WebKit/DerivedSources-input.xcfilelist]: It worked when I had it in alphabetical order as if WK_MODULEMAP_PRIVATE_FILE was expanded, let me try moving it back.

Comment (ian-twilightcoder) [reply Source/WebKit/DerivedSources-input.xcfilelist]: Can't seem to get it to work now 🤔 Oh well, individual paths it is.

File: Source/WebKit/Modules/iOS_Private.modulemap
Comment (ian-twilightcoder) [reply Source/WebKit/Modules/iOS_Private.modulemap]: Confirmed in the build logs for the current version in the OS that `WTF_PLATFORM_MAC` is set for Mac Catalyst builds in the Derived Sources target in WebKit.

I think this is a bug in the makefile. `FEATURE_AND_PLATFORM_DEFINES` is set up using `:=`, and at that point `TARGET_TRIPLE_FLAGS` isn't defined, so it expands to nothing and there's no `-target` argument to clang. TargetConditionals.h will then set `TARGET_OS_OSX`. `SDK_FLAGS` has the same problem so I'll bet `-isysroot` isn't being passed either. I think we get the expected OS/Platform on other platforms due to clang paying attention to the `SDKROOT` environment variable and not really needing `-isysroot`, and synthesizing a more appropriate default target. I'm going to try reordering `TARGET_TRIPLE_FLAGS` and `SDK_FLAGS` and see if that fixes it. Although by fixing that bug, who knows what chaos will ensue...

File: Source/WebKit/DerivedSources-input.xcfilelist
Comment (ian-twilightcoder) [reply Source/WebKit/DerivedSources-input.xcfilelist]: As an aside, why is Derived Sources set up as an aggregate target with a Run Script build phase instead of an External Build System target? It'd save some headache of maintaining an xcfilelist I think?

File: Source/WebKit/Modules/iOS_Private.modulemap
Comment (ian-twilightcoder) [reply Source/WebKit/Modules/iOS_Private.modulemap]: OK I _think_ this should do it finally, running another Mac Catalyst test 🤞

File: Source/WebKit/Configurations/WebKit.xcconfig
Comment (ian-twilightcoder) [reply Source/WebKit/Configurations/WebKit.xcconfig:242]: Diff'ed installhdrs roots and they look good.

**Review by emw-apple — APPROVED** [review]

File: Source/WebKit/Modules/iOS_Private.modulemap
Comment (ian-twilightcoder) [reply Source/WebKit/Modules/iOS_Private.modulemap]: Everything's passing finally! 🥳

**Comment by ian-twilightcoder** [comment]
I still need to run a set of world builds, so don't merge this yet please.

**Comment by ian-twilightcoder** [comment]
> I still need to run a set of world builds, so don't merge this yet please.

This is _finally_ ready to be merged.

**Comment by aproskuryakov** [comment]
The merge queue failure is a pre-existing flaky crash, retrying.

**Pull request merged**
